server/game/GameActions/RevealCards.js

```diff
@@ -0,0 +1,106 @@
+const GameAction = require('./GameAction');
+const HandlerGameActionWrapper = require('./HandlerGameActionWrapper');
+const AcknowledgeRevealCardsPrompt = require('../gamesteps/AcknowledgeRevealCardsPrompt');
+
+class RevealCards extends GameAction {
+    constructor() {
+        super('revealCards');
+        this.defaultWhileRevealed = new HandlerGameActionWrapper({ handler: context => {
+                if(context.revealed.length > 0) {
+                    context.game.queueStep(new AcknowledgeRevealCardsPrompt(context.game, context.revealed, context.player))
+                }
+            }
+        });
+    }
+
+    canChangeGameState({ cards }) {
+        cards = Array.isArray(cards) ? cards : [cards];
+        return cards.length > 0 && cards.some(card => this.isInHiddenArea(card));
+    }
+
+    createEvent({ cards, player, whileRevealed, context }) {
+        const eventParams = {
+            player,
+            cards: Array.isArray(cards) ? cards : [cards],
+            source: context.source
+        }
+        return this.event('onCardsRevealed', eventParams, event => {
+            this.whileRevealed = whileRevealed || this.defaultWhileRevealed;
+            this.playerRevealFunc = (card, player) => event.cards.includes(card) && player === event.player;
+            this.allPlayersRevealFunc = card => event.cards.includes(card);
+            this.revealingTo = context.game.getPlayers();
```

Comment:
Do not add data to the `this` object. GameActions are intended to be stateless; by modifying a `this` property inside the event handler, you are modifying it for ALL `RevealCards` objects.

The `defaultWhileRevealed` in the constructor is fine because it's the same for all RevealCards. But everything else here needs to change.

Reply:
Will be adjusted/fixed in latest commit.

Related to this then, should I also be changing `this.previousSelectCards` (in `highlightRevealedCards` and `hideRevealedCards`)? Not sure of the best way of tracking that otherwise.

Reply:
Anything that uses `this` needs to be changed.

Reply:
Fixed in 4ea24e76c5eb05fa56ddaffcccd34aa31b4856fc and 1155082d26c69f55c0204874c4080920304f301d